File: Source/lua-common/LuaObjectInstanceDescriptor.cpp
//
// Created by 冯鸿杰 on 17/1/13.
//

#include "LuaObjectInstanceDescriptor.h"
#include "LuaObjectClass.h"
#include "LuaNativeClass.hpp"
#include "LuaObjectDecoder.hpp"
#include "LuaObjectEncoder.hpp"
#include "LuaContext.h"

using namespace cn::vimfung::luascriptcore;
using namespace cn::vimfung::luascriptcore::modules::oo;

DECLARE_NATIVE_CLASS(LuaObjectInstanceDescriptor);

LuaObjectInstanceDescriptor::LuaObjectInstanceDescriptor ()
    : LuaObjectDescriptor(), _objectClass (NULL)
{

}

LuaObjectInstanceDescriptor::LuaObjectInstanceDescriptor(const void *instance, LuaObjectClass *objectClass)
    : LuaObjectDescriptor(instance), _objectClass(objectClass)
{
    
}

LuaObjectInstanceDescriptor::LuaObjectInstanceDescriptor (LuaObjectDecoder *decoder)
    : LuaObjectDescriptor(decoder)
{
    std::string clsName = decoder -> readString();
    LuaModule *cls = decoder -> getContext() -> getModule(clsName);
    _objectClass = (LuaObjectClass *)cls;
}

LuaObjectInstanceDescriptor::~LuaObjectInstanceDescriptor()
{
    _objectClass = NULL;
}

std::string LuaObjectInstanceDescriptor::typeName()
{
    static std::string name = typeid(LuaObjectInstanceDescriptor).name();
    return name;
}

LuaObjectClass* LuaObjectInstanceDescriptor::getObjectClass()
{
    return  _objectClass;
}

bool LuaObjectInstanceDescriptor::instanceOf (LuaObjectClass *objectClass)
{
    return this -> getObjectClass() -> subclassOf(objectClass);
}

void LuaObjectInstanceDescriptor::serialization (LuaObjectEncoder *encoder)
{
    LuaObjectDescriptor::serialization(encoder);
    
    //写入类型名称
    encoder -> writeString(this -> getObjectClass() -> getName());
}

void LuaObjectInstanceDescriptor::push(LuaContext *context)
{
    this -> getObjectClass() -> push(this);
}

File: Source/Unity3D/UnityProject/Assets/Plugins/LuaScriptCore/LuaObjectInstanceDescriptor.cs
using UnityEngine;
using System.Collections;
using cn.vimfung.luascriptcore;
using System;
using System.Runtime.InteropServices;

namespace cn.vimfung.luascriptcore.modules.oo
{
	/// <summary>
	/// 实例对象描述器
	/// </summary>
	public class LuaObjectInstanceDescriptor : LuaObjectDescriptor 
	{
		/// <summary>
		/// 初始化
		/// </summary>
		/// <param name="obj">对象</param>
		public LuaObjectInstanceDescriptor(object obj)
			:base(obj)
		{
			
		}

		/// <summary>
		/// 初始化
		/// </summary>
		/// <param name="decode">解码器.</param>
		public LuaObjectInstanceDescriptor(LuaObjectDecoder decode)
			:base(decode)
		{
			
		}

		/// <summary>
		/// 序列化对象
		/// </summary>
		/// <param name="encoder">对象编码器.</param>
		public override void serialization(LuaObjectEncoder encoder)
		{
			base.serialization (encoder);
			encoder.writeString (obj.GetType ().Name);
		}
	}
}
